### PR TITLE
add new property TuneFastDeviceClass

### PR DESCRIPTION
**Pull request by crombus**
Signed-off-by: crombus <pkundra@redhat.com>

### CODE REVIEW

**Comment by crombus** [comment]
@jarrpa I ran `make gen-latest-csv` but crd changes were not done. What can I do to make those changes

**Review by iamniting — CHANGES_REQUESTED** [review]

File: pkg/controller/storagecluster/cephcluster.go [hunk 1]
```diff
@@ -475,3 +484,29 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
+	err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: "", Name: storageClassName}, storageClass)
+	if err != nil {
+		return false, fmt.Errorf("failed to retrieve StorageClass %q. %+v", storageClassName, err)
+	}
+	bool flag := false
+	switch storageClass.Provisioner {
+	case string(EBS):
+		if contains(throttleDiskTypes, storageClass.Parameters["type"]) {
+			flag = true
+			sc.Spec.StorageDeviceSets[i].Config.TuneDeviceClass = true
```
Comment (iamniting) [inline pkg/controller/storagecluster/cephcluster.go:499]: I dont see any variable of SC where did you get it.

File: pkg/controller/storagecluster/cephcluster.go [hunk 2]
```diff
@@ -475,3 +484,29 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
+		}
+	}
+	case string(Premium):
+		if contains(throttleFastDiskTypes, storageClass.Parameters["type"]) {
+			flag = true
+			sc.Spec.StorageDeviceSets[i].Config.TuneFastDeviceClass = true
+		}
+	if flag {
+		sc.Spec.StorageDeviceSets[i].Config.TuneFastDeviceClass = false
+		sc.Spec.StorageDeviceSets[i].Config.TuneDeviceClass = false
```
Comment (iamniting) [inline pkg/controller/storagecluster/cephcluster.go:509]: You need to update SC at the end before return, So these changes will be saved.

**Comment by iamniting** [comment]
I think we should make these changes in two commits first one for the move part and second one for add part.

File: pkg/apis/ocs/v1/storagecluster_types.go [hunk 3]
```diff
@@ -43,7 +43,7 @@ type StorageClusterSpec struct {
 	// If enabled, sets the failureDomain to host, allowing devices to be
 	// distributed evenly across all nodes, regardless of distribution in zones
 	// or racks.
-	FlexibleScaling  bool     `json:"flexibleScaling,omitempty"`
+	FlexibleScaling bool `json:"flexibleScaling,omitempty"`
```
Comment (jarrpa) [inline pkg/apis/ocs/v1/storagecluster_types.go:46]: Move this to its own commit

File: pkg/controller/storagecluster/cephcluster.go [hunk 4]
```diff
@@ -24,6 +24,19 @@ import (
...
 )
 
+// StorageClassProvisionerType is a string representing StorageClass Provisioner. E.g: aws-ebs
+type StorageClassProvisionerType string
+
+// EBS represents AWS EBS provisioner for StorageClass
+EBS StorageClassProvisionerType = "kubernetes.io/aws-ebs"
+
+// Premium represents Azure Premium Managed Disks provisioner for StorageClass
+Premium StorageClassProvisionerType = "kubernetes.io/azure-disk"
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply pkg/controller/storagecluster/cephcluster.go:34]: Agreed. `AzureDisk` would be better.

File: pkg/controller/storagecluster/cephcluster.go [hunk 5]
```diff
@@ -475,3 +484,29 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
 	return resources
 }
+
+func (r *ReconcileStorageCluster) throttleStorageDevices(storageClassName string) (bool, error) {
+	storageClass := &storagev1.StorageClass{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: "", Name: storageClassName}, storageClass)
+	if err != nil {
+		return false, fmt.Errorf("failed to retrieve StorageClass %q. %+v", storageClassName, err)
+	}
+	bool flag := false
```
Comment (jarrpa) [inline pkg/controller/storagecluster/cephcluster.go:494]: `flag` is a terrible name. :) What does this flag represent? Also you probably don't need `bool` at the front.

File: pkg/controller/storagecluster/cephcluster.go (hunk 1 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/cephcluster.go:499]:
```suggestion
			sc.Spec.StorageDeviceSets[i].Config.TuneFastDeviceClass = true
```

File: pkg/controller/storagecluster/cephcluster.go (hunk 2 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/cephcluster.go:509]:
```suggestion
		sc.Spec.StorageDeviceSets[i].Config.TuneFastDeviceClass = false
```

File: pkg/controller/storagecluster/cephcluster.go [hunk 6]
```diff
@@ -475,3 +484,29 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
...
+	switch storageClass.Provisioner {
```
Comment (jarrpa) [inline pkg/controller/storagecluster/cephcluster.go:495]: This entire switch statement looks broken. Since you've removed the WIP tag, you should be running `make ocs-operator-ci` before ever pushing to GitHub to catch things like this.

**Review by jarrpa — CHANGES_REQUESTED** [review]

File: pkg/controller/storagecluster/cephcluster.go (hunk 2 above)
Comment (nbalacha) [reply pkg/controller/storagecluster/cephcluster.go:509]: > You need to update SC at the end before return, So these changes will be saved.

And would be better if done in ensureCephCluster as updating it in this function could mean possibly updating the storageCluster for each storageDeviceSet.

File: pkg/controller/storagecluster/cephcluster.go [hunk 7]
```diff
@@ -475,3 +484,29 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
 
 	return resources
 }
+
+func (r *ReconcileStorageCluster) throttleStorageDevices(storageClassName string) (bool, error) {
```
Comment (nbalacha) [inline pkg/controller/storagecluster/cephcluster.go:488]: The previous function looks better to me - it just determines whether the storageClass should be throttled and lets the calling function handle the rest.

**Comment by crombus** [comment]
@jarrpa the ci test is failing at govet. need help there.

**Review by kesavanvt — APPROVED** [review]

**Comment by kesavanvt** [comment]
@crombus since the implementation of `throttleStorageDevices` method is modified, it's necessary to make necessary changes in `TestThrottleStorageDevices` (`throttleStorageDevices` is invoked here ) under `storagecluster_controller_test.go`.

File: pkg/controller/storagecluster/cephcluster.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (crombus) [reply pkg/controller/storagecluster/cephcluster.go]: can you please refresh. I have done these changes.

[the remark this answers is not in the document]
Comment (kesavanvt) [reply pkg/controller/storagecluster/cephcluster.go]: Yes, i have approved it.

[the remark this answers is not in the document]
Comment (crombus) [reply pkg/controller/storagecluster/cephcluster.go]: cool thanks

File: pkg/controller/storagecluster/cephcluster.go [hunk 8]
```diff
@@ -41,15 +55,12 @@ func (r *ReconcileStorageCluster) ensureCephCluster(sc *ocsv1.StorageCluster, re
 	// this is for performance optimization of slow device class
 	//TODO: If for a StorageDeviceSet there is a separate metadata pvc template, check for StorageClass of data pvc template only
 	for i, ds := range sc.Spec.StorageDeviceSets {
-		throttle, err := r.throttleStorageDevices(*ds.DataPVCTemplate.Spec.StorageClassName)
+		sd, fd, err := r.throttleStorageDevices(*ds.DataPVCTemplate.Spec.StorageClassName)
```
Comment (jarrpa) [inline pkg/controller/storagecluster/cephcluster.go:58]:
```suggestion
		throttleSlow, throttleFast, err := r.throttleStorageDevices(*ds.DataPVCTemplate.Spec.StorageClassName)
```

File: pkg/controller/storagecluster/cephcluster.go [hunk 9]
```diff
@@ -24,11 +25,24 @@ import (
...
 	// Hardcoding networkProvider to multus and this can be changed later to accomodate other providers
 	networkProvider           = "multus"
 	publicNetworkSelectorKey  = "public"
 	clusterNetworkSelectorKey = "cluster"
+
+	// AzureDisk represents Azure Premium Managed Disks provisioner for StorageClass
+	AzureDisk StorageClassProvisionerType = "kubernetes.io/azure-disk"
+
+	// EBS represents AWS EBS provisioner for StorageClass
+	EBS StorageClassProvisionerType = "kubernetes.io/aws-ebs"
```
Comment (jarrpa) [inline pkg/controller/storagecluster/cephcluster.go:45]: These should remain in `reconcile.go`.

**Review by jarrpa — CHANGES_REQUESTED** [review]

**Comment by jarrpa** [comment]
Test failures are valid. Please rebase and resolve. **Always** run `make ocs-operator-ci` before any push.

**Comment by crombus** [comment]
the PR is dependent on https://github.com/openshift/ocs-operator/pull/912. rook 1.5.1. I'll rebase it.
The failures are because of govet.

**Comment by obnoxxx** [comment]
> the PR is dependent on #912. rook 1.5.1. I'll rebase it.
> The failures are because of govet.

#912 has merged now @crombus - you can rebase

**Comment by obnoxxx** [comment]
Not strictly required for a 4.7 feature, afaict, but would be good to have before feature freeze...

File: pkg/controller/storagecluster/reconcile.go [hunk 10]
```diff
@@ -71,6 +69,12 @@ osd_memory_target_cgroup_limit_ratio = 0.5
 
 	// DeviceTypeNVMe represents the DeviceType NVMe
 	DeviceTypeNVMe = "nvme"
+
+	// AzureDisk represents Azure Premium Managed Disks provisioner for StorageClass
+        AzureDisk StorageClassProvisionerType = "kubernetes.io/azure-disk"
+        
+        // EBS represents AWS EBS provisioner for StorageClass
+        EBS StorageClassProvisionerType = "kubernetes.io/aws-ebs"
```
Comment (jarrpa) [inline pkg/controller/storagecluster/reconcile.go:77]: Fix indentation here.

File: pkg/controller/storagecluster/storagecluster_controller_test.go [hunk 11]
```diff
@@ -260,7 +260,8 @@ func TestThrottleStorageDevices(t *testing.T) {
 		label          string
 		storageClass   *storagev1.StorageClass
 		storageCluster *api.StorageCluster
-		expected       bool
+		expectedSlow   bool
+                expectedFast   bool
```
Comment (kesavanvt) [inline pkg/controller/storagecluster/storagecluster_controller_test.go:264]: Please fix the indentation and formatting, you can do it by `go fmt <go-file>`

Comment (kesavanvt) [reply pkg/controller/storagecluster/storagecluster_controller_test.go:256]: Still the indentation and formatting seems to be incorrect for this go file.

**Review by kesavanvt — APPROVED** [review]

File: pkg/controller/storagecluster/cephcluster.go [hunk 12]
```diff
@@ -184,8 +186,8 @@ func newCephCluster(sc *ocsv1.StorageCluster, cephImage string, nodeCount int, s
 			},
 			Mgr: cephv1.MgrSpec{
 				Modules: []cephv1.Module{
-					cephv1.Module{Name: "pg_autoscaler", Enabled: true},
-					cephv1.Module{Name: "balancer", Enabled: true},
+					{Name: "pg_autoscaler", Enabled: true},
+					{Name: "balancer", Enabled: true},
```
Comment (jarrpa) [inline pkg/controller/storagecluster/cephcluster.go:190]: These changes are unrelated and should be either removed or in their own commit.

File: pkg/controller/storagecluster/reconcile.go [hunk 13]
```diff
@@ -26,13 +25,13 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
+// StorageClassProvisionerType is a string representing StorageClass Provisioner. E.g: aws-ebs
+type StorageClassProvisionerType string
+
```
Comment (jarrpa) [inline pkg/controller/storagecluster/reconcile.go:30]: This reorder is uneccessary, please move it back.

File: pkg/controller/storagecluster/reconcile.go [hunk 14]
```diff
@@ -424,7 +426,7 @@ func (r *ReconcileStorageCluster) validateStorageDeviceSets(sc *ocsv1.StorageClu
 			}
 		}
 		if ds.DeviceType != "" {
-			if ( DeviceTypeSSD == strings.ToLower(ds.DeviceType )) || ( DeviceTypeHDD == strings.ToLower(ds.DeviceType )) || ( DeviceTypeNVMe == strings.ToLower(ds.DeviceType )) {
+			if (DeviceTypeSSD == strings.ToLower(ds.DeviceType)) || (DeviceTypeHDD == strings.ToLower(ds.DeviceType)) || (DeviceTypeNVMe == strings.ToLower(ds.DeviceType)) {
```
Comment (jarrpa) [inline pkg/controller/storagecluster/reconcile.go:429]: This formatting change is being taken care of in a different PR, please remove it.

File: pkg/controller/storagecluster/storagecluster_controller_test.go [hunk 15]
```diff
@@ -125,7 +125,7 @@ var mockNodeList = &corev1.NodeList{
 		Kind: "NodeList",
 	},
 	Items: []corev1.Node{
-		corev1.Node{
+		{
```
Comment (jarrpa) [inline pkg/controller/storagecluster/storagecluster_controller_test.go:128]: All changes like this are unrelated formatting changes, please move to their own commit.

**Review by jarrpa — CHANGES_REQUESTED** [review]
For the test commit, it needs to be split into two:

* Any changes that are simply fixing the existing tests due to your code changes should go with your code commit so that the commit doesn't break the tests
* The new test case should be added in its own commit.

File: pkg/controller/storagecluster/cephcluster.go [hunk 16]
```diff
@@ -24,6 +24,19 @@ import (
...
+
+// EBS represents AWS EBS provisioner for StorageClass
+EBS StorageClassProvisionerType = "kubernetes.io/aws-ebs"
+
+// Premium represents Azure Premium Managed Disks provisioner for StorageClass
+Premium StorageClassProvisionerType = "kubernetes.io/azure-disk"
+
+var throttleDiskTypes = []string{"gp2", "io1"}
+
+var throttleFastDiskTypes =[]string{"managed-permium"}
```
[the remark this answers is not in the document]
Comment (obnoxxx) [reply pkg/controller/storagecluster/cephcluster.go]: done

File: pkg/controller/storagecluster/cephcluster.go [hunk 17]
```diff
@@ -41,15 +54,10 @@ func (r *ReconcileStorageCluster) ensureCephCluster(sc *ocsv1.StorageCluster, re
 	// this is for performance optimization of slow device class
 	//TODO: If for a StorageDeviceSet there is a separate metadata pvc template, check for StorageClass of data pvc template only
 	for i, ds := range sc.Spec.StorageDeviceSets {
-		throttle, err := r.throttleStorageDevices(*ds.DataPVCTemplate.Spec.StorageClassName)
+		err := r.throttleStorageDevices(*ds.DataPVCTemplate.Spec.StorageClassName)
```
[the remark this answers is not in the document]
Comment (obnoxxx) [reply pkg/controller/storagecluster/cephcluster.go]: this is done

File: pkg/controller/storagecluster/cephcluster.go (hunk 7 above)
Comment (obnoxxx) [reply pkg/controller/storagecluster/cephcluster.go]: this has been taken care of, afaict

**Review by obnoxxx — APPROVED** [review]
To my taste, the main commit (add new property TuneFastDeviceClass) does a bit too much. I would have recommended to split the refactoring (moving of function from `reconcile.go` to `cephcluster.go`, etc) into separat commits. But not blocking because of this.

The whole terminology with `Tune[Fast|Slow]DeviceClass` is rather confusing or non-obvious, but I guess it is what it is.

For now the patch might be enough, but conceptually, I think we need the list of device types to tune for slow or fast types not to be just storage class names and hard code selection of provisioners in the throttle function but instead we should have the devices types be pairs of provisioner and storage class name.

Since this can be done in a follow-up cleanup PR, approving this one for now.

/lgtm

**Comment by agarwal-mudit** [comment]
@crombus https://bugzilla.redhat.com/show_bug.cgi?id=1909793 still doesn't have all the acks. Pls do not merge it into 4.6 yet.
Also, you should first backport it to 4.7

**Comment by crombus** [comment]
@agarwal-mudit this feature is already in 4.7.
And this will require manual BP of the two PR's mentioned in BZ.
I would say it is tricky to BP.

**Comment by agarwal-mudit** [comment]
> @agarwal-mudit this feature is already in 4.7.
> And this will require manual BP of the two PR's mentioned in BZ.
> I would say it is tricky to BP.

Ok, then lets not backport it to 4.6
Probably we can GA Azure without this feature. @obnoxxx
